keps/sig-node/127-user-namespaces/README.md

```diff
@@ -316,121 +327,108 @@ The picked range will be stored under a file named `userns` in the pod folder
-listed vulnerabilities (as the host is protected from the container). It is also
-a trivial next-step to take, given that we have phase 1 implemented: just return
-the same mapping if the pod has other volumes.
+Now, when trying to read the file the pod created with UID 65536 on its on disk
+inode, it will fail. Host UID 65536 is not mapped, therefore it can't read it.
 
-While these pods do not use a distinct user namespace mapping, they are still
-using a new user namespace object in the kernel (so they cannot join/attack
-other pods namespaces). Security-wise this is a middle layer between what we
-have today (no userns at all) and using a distinct UID/GID mapping for the user
-namespace.
+A similar problem happens if a pod wants to share files with other pods that run
+with different mappings.
 
-#### Phase 3: TBD
+##### Example without idmap mounts
 
-#### Unresolved
+Now let's say the pod is using its volumes that were mounted using idmap mounts
+by the container runtime. All the mappings used (the idmap mounts and the pod
+userns) are:
 
-Here is a list of considerations raised in PRs discussion that hasn't yet
-settle. This list is not exhaustive, we are just trying to put the things that
-we don't want to forget or want to highlight it. Some things that are obvious we
-need to tackle are not listed. Let us know if you think it is important to add
-something else to this list:
+| Host ID  | Container ID   | Length  |
+|----------|----------------|---------|
+| 65536    |      0         |  65536  |
+
+
+Now, when the pod running as UID 0 creates a file in the volume, the on disk UID
+of the file will be 0. This means, when writing via the idmap mount, the kernel
+does an ID translation and the UID inside the container is used for the file on
+disk inode. Something similar happens when reading the file: the UID 0 the file
+has in its inode, is allowed for the pod to read it.
 
-- We will start with mappings of 64K. Tim Hockin, however, has expressed
-  concerns. See more info on [this Github discussion](https://github.com/kubernetes/enhancements/pull/3065#discussion_r781676224)
-  SergeyKanzhelev [suggested a nice alternative](https://github.com/kubernetes/enhancements/pull/3065#discussion_r807408134),
-  to limit the number of pods so we guarantee enough spare host UIDs in case we
-  need them for the future. There is no final decision yet on how to handle this.
-  For now we will limit the number of pods, so the wide mapping is not
-  problematic, but [there are downsides to this too](https://github.com/kubernetes/enhancements/pull/3065#discussion_r812806223)
+This solves the problems we had before, in the example without idmap mounts. If
+the pod is re-scheduled with a different mapping, reading the volume files will
+not be an issue as long as it is running with the same ID inside the container.
 
+For that same reason, sharing volumes with other pods also work, as long as they
```

Comment:
I'm confused - isn't one of the goals of this whole effort to PREVENT the ability for other pods to read my stuff?

Consider an attacker that finds a way to escape the pivot_root and read the root filesystem.  Can they navigate to another Pod's kubelet dir and read that pod's on-disk data?

When we spent so much time figuring out how to do this safely, one of the primary constraints was that no two namespaces (or exceedingly rarely) would use the same ID mapping, ensuring that even if you escaped your root, you could still count on the OS's long-proven UID/GID handling.

What am I missing?

Reply:
So, first, brain fart here, this example should not be in the stateless pods KEP :-D.

No, an attacker that escapes pivot root and can read the filesystem, won't be able to read other's pod on-disk data. The point you are missing, that maybe wasn't clear in my example that shouldn't be in this KEP, is that this IDs transformation is done only on this bind mount. IOW, if you do bind-mount with idmap of `source` on `target` only when you access via the `target` mount point you will have this effect.

So no, in NO way all interactions with the system have this behavior. This behavior is only when you access a idmap mount (that, btw, can't be created inside a userns, so the container won't be able to create idmap mounts either)